src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs

```diff
@@ -0,0 +1,884 @@
+            (CharToBase32[base32[3]] << 6) | (CharToBase32[base32[4]] << 1) | (CharToBase32[base32[5]] >> 4)
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 3) = (byte)(
+            (CharToBase32[base32[5]] << 4) | (CharToBase32[base32[6]] >> 1)
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 4) = (byte)(
+            (CharToBase32[base32[6]] << 7) | (CharToBase32[base32[7]] << 2) | (CharToBase32[base32[8]] >> 3)
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 5) = (byte)(
+            (CharToBase32[base32[8]] << 5) | CharToBase32[base32[9]]
+        );
+
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 6) = (byte)(
+            (CharToBase32[base32[10]] << 3) | (CharToBase32[base32[11]] >> 2)
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 7) = (byte)(
+            (CharToBase32[base32[11]] << 6) | (CharToBase32[base32[12]] << 1) | (CharToBase32[base32[13]] >> 4)
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 8) = (byte)(
+            (CharToBase32[base32[13]] << 4) | (CharToBase32[base32[14]] >> 1)
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 9) = (byte)(
+            (CharToBase32[base32[14]] << 7) | (CharToBase32[base32[15]] << 2) | (CharToBase32[base32[16]] >> 3)
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 10) = (byte)(
+            (CharToBase32[base32[16]] << 5) | CharToBase32[base32[17]]
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 11) = (byte)(
+            (CharToBase32[base32[18]] << 3) | CharToBase32[base32[19]] >> 2
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 12) = (byte)(
+            (CharToBase32[base32[19]] << 6) | (CharToBase32[base32[20]] << 1) | (CharToBase32[base32[21]] >> 4)
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 13) = (byte)(
+            (CharToBase32[base32[21]] << 4) | (CharToBase32[base32[22]] >> 1)
+        );
+        Unsafe.Add(ref Unsafe.As<Ulid, byte>(ref ulid), 14) = (byte)(
+            (CharToBase32[base32[22]] << 7) | (CharToBase32[base32[23]] << 2) | (CharToBase32[base32[24]] >> 3)
+        );
+
+        return ulid;
+    }
+
+    // Convert
+
+    public byte[] ToByteArray()
+    {
+        var bytes = new byte[16];
+        Unsafe.WriteUnaligned(ref bytes[0], this);
+        return bytes;
+    }
+
+    public bool TryWriteBytes(Span<byte> destination)
+    {
+        if (destination.Length < 16)
+        {
+            return false;
+        }
+
+        Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(destination), this);
+        return true;
+    }
+
+    public string ToBase64(Base64FormattingOptions options = Base64FormattingOptions.None)
+    {
+        var buffer = ArrayPool<byte>.Shared.Rent(16);
+        try
+        {
+            TryWriteBytes(buffer);
+            return Convert.ToBase64String(buffer, options);
```

Comment:
`Convert.ToBase64String` has an overload that accepts `ROS<byte>` so you could avoid this by stackalloc'ing 16 bytes, writing there and converting that.